cleo_sdk/CLEO_Utils.h

```diff
@@ -0,0 +1,427 @@
+        vsprintf(szBuf, format, args); // put params into format
+        CLEO_Log(level, szBuf);
+        return szBuf;
+    }
+
+    static void Trace(CLEO::eLogLevel level, const char* format, ...)
+    {
+        va_list args;
+        va_start(args, format);
+        TraceVArg(level, format, args);
+        va_end(args);
+    }
+
+    static void Trace(const CLEO::CRunningScript* thread, CLEO::eLogLevel level, const char* format, ...)
+    {
+        if (thread != nullptr && CLEO_GetScriptVersion(thread) < CLEO::eCLEO_Version::CLEO_VER_5)
+        {
+            return; // do not log this in older versions
+        }
+
+        va_list args;
+        va_start(args, format);
+        TraceVArg(level, format, args);
+        va_end(args);
+    }
+
+    static void ShowError(const char* format, ...)
+    {
+        va_list args;
+        va_start(args, format);
+        auto msg = TraceVArg(CLEO::eLogLevel::Error, format, args);
+        va_end(args);
+
+        QUERY_USER_NOTIFICATION_STATE pquns;
+        SHQueryUserNotificationState(&pquns);
+        bool fullscreen = (pquns == QUNS_BUSY) || (pquns == QUNS_RUNNING_D3D_FULL_SCREEN) || (pquns == QUNS_PRESENTATION_MODE);
+
+        if (fullscreen)
+        {
+            PostMessage(NULL, WM_SYSCOMMAND, SC_MINIMIZE, 0);
+            ShowWindow(NULL, SW_MINIMIZE);
+        }
+
+        MessageBox(NULL, msg, "CLEO error", MB_SYSTEMMODAL | MB_TOPMOST | MB_ICONERROR | MB_OK);
+
+        if (fullscreen)
+        {
+            PostMessage(NULL, WM_SYSCOMMAND, SC_RESTORE, 0);
+            ShowWindow(NULL, SW_RESTORE);
+        }
+    }
+
+    #define TRACE(format,...) {CLEO::Trace(CLEO::eLogLevel::Default, format, __VA_ARGS__);}
+    #define LOG_WARNING(script, format, ...) {CLEO::Trace(script, CLEO::eLogLevel::Error, format, __VA_ARGS__);}
+    #define SHOW_ERROR(a,...) {CLEO::ShowError(a, __VA_ARGS__);}
+
+    const size_t MinValidAddress = 0x10000; // used for validation of pointers received from scripts. First 64kb are for sure reserved by Windows.
+    #define OPCODE_VALIDATE_POINTER(x) if((size_t)x <= MinValidAddress) { SHOW_ERROR("Invalid '0x%X' pointer argument in script %s \nScript suspended.", x, ScriptInfoStr(thread).c_str()); return thread->Suspend(); }
+
+    #define OPCODE_CONDITION_RESULT(value) CLEO_SetThreadCondResult(thread, value);
+
+    // opcode param handling utils internal
+    static SCRIPT_VAR* _paramsArray = nullptr;
+    static eDataType _lastParamType = eDataType::DT_END;
+    static eArrayDataType _lastParamArrayType = eArrayDataType::ADT_NONE;
+
+    static SCRIPT_VAR& _getParamSimple(CRunningScript* thread)
+    {
+        _lastParamType = thread->PeekDataType();
+        _lastParamArrayType = IsArray(_lastParamType) ? thread->PeekArrayDataType() : eArrayDataType::ADT_NONE;
+
+        CLEO_RetrieveOpcodeParams(thread, 1);
+        if (_paramsArray == nullptr) _paramsArray = CLEO_GetParamsCollectiveArray();
+        return _paramsArray[0];
+    }
```

Comment:
can you add a comment on what this method does? It is unclear what is "simple"

Reply:
maybe call it `_ReadNextParam`?

Reply:
updated